src/transformers/trainer.py

```diff
@@ -519,8 +522,22 @@ def __init__(
                 "Passing `optimizers` is not allowed if Deepspeed or PyTorch FSDP is enabled. "
                 "You should subclass `Trainer` and override the `create_optimizer_and_scheduler` method."
             )
+
+        self.chkpt_manager = None
+        if self.args.use_checkpoint_manager:
+            # TODO(jonbolin): Currently we manually initialize the process
+            # group in SPMD mode. This should be moved behind Accelerate.
+            torch.distributed.init_process_group('gloo', init_method='xla://')
+            self.chkpt_manager = CheckpointManager(
+                path=self.args.output_dir,
+                save_interval=float('inf'),  # Defer to the HF save strategy
+                max_to_keep=self.args.save_total_limit or 0
+            )
+
         default_callbacks = DEFAULT_CALLBACKS + get_reporting_integration_callbacks(self.args.report_to)
         callbacks = default_callbacks if callbacks is None else default_callbacks + callbacks
+        if self.args.use_checkpoint_manager:
+            callbacks.append(CheckpointManagerCallback(self.chkpt_manager))
```

Comment:
Can you tell me more about what these callbacks do?

Reply:
These callbacks are invoked at various times in training, e.g. `on_step_end`, `on_train_start`, `on_epoch_end`. They're meant to modify the control state or take action at certain times, which for us would mean modifying `control.should_save` based on `chpkt_manager.should_save()` each step (`on_step_end`).

Since the HF control manages checkpoint frequency, the only time `chpkt_manager.should_save` would override the value is when a preemption is caught.

Reply:
Thanks for the explanation.